OpenApiDataObjectScanner throws ArrayIndexOutOfBoundsException while mapping a generic with 2 Parameters

Hi,
I am trying to return [io.vavr.control.Either<L,R>](https://www.javadoc.io/doc/io.vavr/vavr/0.9.2/io/vavr/control/Either.html) in a REST endpoint on Quarkus. Class Either<L,R> has an iterable super interface. Thus, OpenApiDataObjectScanner picks Iterable<E>. That causes a mismatch and process fails with java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1. 
TIA...

Version: Quarkus 2.2.2.Final
Dependency: io.quarkus:quarkus-smallrye-openapi
Build tool: Gradle

`Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
	at java.base/java.util.Arrays$ArrayList.get(Arrays.java:4164)
	at java.base/java.util.Collections$UnmodifiableList.get(Collections.java:1321)
	at io.smallrye.openapi.runtime.scanner.dataobject.TypeResolver.buildParamTypeResolutionMap(TypeResolver.java:1025)
	at io.smallrye.openapi.runtime.scanner.dataobject.TypeResolver.getAllFields(TypeResolver.java:460)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.resolveSpecial(OpenApiDataObjectScanner.java:319)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:206)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:157)
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.otherTypeToSchema(SchemaFactory.java:582)
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.typeToSchema(SchemaFactory.java:428)
	at io.smallrye.openapi.runtime.scanner.spi.AnnotationScanner.createResponseFromRestMethod(AnnotationScanner.java:414)
	at io.smallrye.openapi.runtime.scanner.spi.AnnotationScanner.processResponse(AnnotationScanner.java:346)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceMethod(JaxRsAnnotationScanner.java:435)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.lambda$processResourceMethods$0(JaxRsAnnotationScanner.java:271)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceMethods(JaxRsAnnotationScanner.java:269)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceClass(JaxRsAnnotationScanner.java:239)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceClasses(JaxRsAnnotationScanner.java:214)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.scan(JaxRsAnnotationScanner.java:146)
	at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.scan(OpenApiAnnotationScanner.java:127)
	at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.generateAnnotationModel(SmallRyeOpenApiProcessor.java:452)
	at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.build(SmallRyeOpenApiProcessor.java:320)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:832)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)`

--- a/core/src/main/java/io/smallrye/openapi/runtime/scanner/OpenApiDataObjectScanner.java
+++ b/core/src/main/java/io/smallrye/openapi/runtime/scanner/OpenApiDataObjectScanner.java
@@ -18,8 +18,10 @@ import org.jboss.jandex.DotName;
 import org.jboss.jandex.Index;
 import org.jboss.jandex.IndexView;
 import org.jboss.jandex.Indexer;
+import org.jboss.jandex.ParameterizedType;
 import org.jboss.jandex.PrimitiveType;
 import org.jboss.jandex.Type;
+import org.jboss.jandex.Type.Kind;
 
 import io.smallrye.openapi.api.models.media.SchemaImpl;
 import io.smallrye.openapi.api.models.media.XMLImpl;
@@ -316,9 +318,31 @@ public class OpenApiDataObjectScanner {
     }
 
     private void resolveSpecial(DataObjectDeque.PathEntry root, Type type) {
+        if (typeArgumentMismatch(type, rootClassInfo)) {
+            /*
+             * The type's generic arguments are not in alignment with the type
+             * parameters of the stand-in collection type. For the purposes
+             * of obtaining the stand-in collection's schema, we discard the
+             * original type and determine the correct type parameter for the
+             * stand-in.
+             */
+            ParameterizedType standinInterface = rootClassInfo.interfaceTypes().get(0).asParameterizedType();
+            type = TypeResolver.resolveParameterizedAncestor(context,
+                    type.asParameterizedType(),
+                    standinInterface);
+        }
+
         Map<String, TypeResolver> fieldResolution = TypeResolver.getAllFields(context, type, rootClassInfo,
                 root.getAnnotationTarget());
         rootSchema = preProcessSpecial(type, fieldResolution.values().iterator().next(), root);
+    }
+
+    private boolean typeArgumentMismatch(Type type, ClassInfo standin) {
+        if (type.kind() != Kind.PARAMETERIZED_TYPE) {
+            return false;
+        }
+
+        return standin.typeParameters().size() < type.asParameterizedType().arguments().size();
     }
 
     private Schema preProcessSpecial(Type type, TypeResolver typeResolver, DataObjectDeque.PathEntry currentPathEntry) {

--- a/extension-jaxrs/src/test/java/io/smallrye/openapi/runtime/scanner/ApiResponseTests.java
+++ b/extension-jaxrs/src/test/java/io/smallrye/openapi/runtime/scanner/ApiResponseTests.java
@@ -3,6 +3,7 @@ package io.smallrye.openapi.runtime.scanner;
 import java.io.IOException;
 import java.util.Collection;
 import java.util.HashMap;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.NavigableMap;
@@ -275,4 +276,28 @@ class ApiResponseTests extends IndexScannerTestBase {
                 kotlin.coroutines.CoroutineContext.class);
     }
 
+    @Test
+    void testStandinGenericTypeArgumentResolution() throws IOException, JSONException {
+        abstract class Either<L, R> implements Iterable<R> {
+            @Override
+            public Iterator<R> iterator() {
+                return null;
+            }
+        }
+
+        @jakarta.ws.rs.Path("/")
+        class Test {
+            @jakarta.ws.rs.GET
+            @jakarta.ws.rs.Path("/either")
+            @jakarta.ws.rs.Produces({ "text/plain" })
+            public Either<Integer, String> getEither() {
+                return null;
+            }
+        }
+        Index index = indexOf(Test.class, Either.class);
+        OpenApiAnnotationScanner scanner = new OpenApiAnnotationScanner(emptyConfig(), index);
+        OpenAPI result = scanner.scan();
+        printToConsole(result);
+        assertJsonEquals("responses.standin-generic-type-resolved.json", result);
+    }
 }

--- a/extension-jaxrs/src/test/resources/io/smallrye/openapi/runtime/scanner/responses.standin-generic-type-resolved.json
+++ b/extension-jaxrs/src/test/resources/io/smallrye/openapi/runtime/scanner/responses.standin-generic-type-resolved.json
@@ -1,0 +1,31 @@
+{
+  "openapi": "3.0.3",
+  "paths": {
+    "/either": {
+      "get": {
+        "responses": {
+          "200": {
+            "description": "OK",
+            "content": {
+              "text/plain": {
+                "schema": {
+                  "$ref": "#/components/schemas/1EitherIntegerString"
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  },
+  "components": {
+    "schemas": {
+      "1EitherIntegerString": {
+        "type": "array",
+        "items": {
+          "type": "string"
+        }
+      }
+    }
+  }
+}

